resolwe_runtime_utils.py

```diff
@@ -15,25 +15,114 @@
 """
 Utility functions that make it easier to write a Resolwe process.
 """
+import functools
 import glob
 import gzip
 import json
+import logging
 import os
 import re
 import shlex
 import shutil
+import socket
 import subprocess
 import tarfile
 import zlib
+import time
+from contextlib import suppress
+from itertools import chain
+from pathlib import Path
 
-import requests
+
+# Socket constants. The timeout is set to infinite.
+SOCKET_TIMEOUT = None
+if "SOCKET_TIMEOUT" is os.environ:
+    SOCKET_TIMEOUT = int(os.environ["SOCKET_TIMEOUT"])
+SOCKETS_PATH = Path(os.environ.get("SOCKETS_VOLUME", "/sockets"))
+COMMUNICATOR_SOCKET = SOCKETS_PATH / os.environ.get("SCRIPT_SOCKET", "_socket2.s")
+DATA_VOLUME = Path(os.environ.get("DATA_VOLUME", "/data"))
+
+logger = logging.getLogger(__name__)
 
 
 # Compat between Python 2.7/3.4 and Python 3.5
 if not hasattr(json, 'JSONDecodeError'):
     json.JSONDecodeError = ValueError
 
 
+def _retry(
+    max_retries=3,
+    retry_exceptions=(ConnectionError, FileNotFoundError),
+    min_sleep=1,
+    max_sleep=10,
+):
+    """Try to call decorated method max_retries times before giving up.
+
+    The calls are retried when function raises exception in retry_exceptions.
+
+    :param max_retries: maximal number of calls before giving up.
+    :param retry_exceptions: retry call if one of these exceptions is raised.
+    :param min_sleep: minimal sleep between calls (in seconds).
+    :param max_sleep: maximal sleep between calls (in seconds).
+    :returns: return value of the called method.
+    :raises: the last exceptions raised by the method call if none of the
+      retries were successfull.
+    """
+
+    def decorator_retry(func):
+        @functools.wraps(func)
+        def wrapper_retry(*args, **kwargs):
+            last_error = Exception("Retry failed")
+            sleep = 0
+            for retry in range(max_retries):
+                try:
+                    time.sleep(sleep)
+                    return func(*args, **kwargs)
+                except retry_exceptions as err:
+                    sleep = min(max_sleep, min_sleep * (2 ** retry))
+                    last_error = err
+            raise last_error
+
+        return wrapper_retry
+
+    return decorator_retry
+
+
+def _read_bytes(s, message_size):
```

Comment:
I don't like single-letter variables. What about naming it `socket`?

Reply:
Socket shadows the socket module. I have renamed it to `socket_`.